File: templates/problems/11_graphs/11_num_components/go/num_components_test.go
package num_components

import "testing"

func TestCountComponents(t *testing.T) {
	tests := []struct {
		name  string
		n     int
		edges [][]int
		want  int
	}{
		{
			name:  "two components",
			n:     5,
			edges: [][]int{{0, 1}, {1, 2}, {3, 4}},
			want:  2,
		},
		{
			name:  "one component",
			n:     5,
			edges: [][]int{{0, 1}, {1, 2}, {2, 3}, {3, 4}},
			want:  1,
		},
		{
			name:  "all isolated",
			n:     4,
			edges: [][]int{},
			want:  4,
		},
		{
			name:  "single node",
			n:     1,
			edges: [][]int{},
			want:  1,
		},
		{
			name:  "three components",
			n:     6,
			edges: [][]int{{0, 1}, {2, 3}, {4, 5}},
			want:  3,
		},
		{
			name:  "fully connected",
			n:     3,
			edges: [][]int{{0, 1}, {1, 2}, {0, 2}},
			want:  1,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := countComponents(tt.n, tt.edges)
			if got != tt.want {
				t.Errorf("countComponents(%d, %v) = %v, want %v", tt.n, tt.edges, got, tt.want)
			}
		})
	}
}

File: templates/problems/11_graphs/11_num_components/typescript/num_components.test.ts
import { describe, it, expect } from "vitest";
import { countComponents } from "./num_components";

describe("countComponents", () => {
    it.each([
        { name: "two components", n: 5, edges: [[0, 1], [1, 2], [3, 4]], want: 2 },
        { name: "one component", n: 5, edges: [[0, 1], [1, 2], [2, 3], [3, 4]], want: 1 },
        { name: "all isolated", n: 4, edges: [], want: 4 },
        { name: "single node", n: 1, edges: [], want: 1 },
        { name: "three components", n: 6, edges: [[0, 1], [2, 3], [4, 5]], want: 3 },
        { name: "fully connected", n: 3, edges: [[0, 1], [1, 2], [0, 2]], want: 1 },
    ])("case: $name", ({ n, edges, want }) => {
        const got = countComponents(n, edges);
        expect(got).toBe(want);
    });
});
